### DISCUSSION TITLE
How to translate GitFourchette to other languages?

### DISCUSSION THREAD

**Discussion by Flaconia** [jorio/gitfourchette#5] [Q&A] [answered]
I'm loving this client so much, I already know it's going to be my only client on Linux 🐧

I'd like to add an Italian localization. Can anybody indicate me how to do it?

**Comment by jorio** [comment]
Hey! Glad you like GitFourchette :) An Italian localization would be great! [Here's a localization guide.](https://github.com/jorio/gitfourchette/blob/master/gitfourchette/assets/lang/README.md)

If you need to get started with a blank `.ts` file for Italian, I've generated one for you here: [gitfourchette_it.ts.gz](https://github.com/user-attachments/files/18118865/gitfourchette_it.ts.gz). Extract it and put it in `gitfourchette/assets/lang` beside the existing .ts files for English and French. Then, open it in Qt Linguist and start translating.

The guide explains the rest. You may need PySide6 tool wrappers to use some of the commands in the guide, such as `pyside6-lupdate` – these should be available via `pip install PySide6`.

Let me know if anything's unclear or if you need help. Thanks!

**Reply by Flaconia** [reply]
Hey thanks to you for your guidance!
I managed to set up almost everything. Cloned the repo, installed Qt Linguist and did all you mentioned to start translating.
I've got a problem when running the py scripts though;

>  The guide explains the rest. You may need PySide6 tool wrappers to use some of the commands in the guide, such as `pyside6-lupdate` – these should be available via `pip install PySide6`.

First of all, I'm on Manjaro. In Pamac Manager, both pyside6 and pyside6-tools show up as installed, but when running the update script, looks like something's missing:

```
pyside_tool.py -source-language en -target-language it \  
> - extensions py,ui ./gitfourchette/ \
> ts gitfourchette/assets/lang/gitfourchette.it.ts
Traceback (most recent call last):
  File "/usr/bin/pyside_tool.py", line 249, in <module>
    main()
  File "/usr/bin/pyside_tool.py", line 47, in main
    sys.exit(subprocess.call(command))
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/subprocess.py", line 389, in call
    with Popen(*popenargs, **kwargs) as p:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.12/subprocess.py", line 1955, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/usr/bin/../pyside_tool.py'

```

I've also tried running the script directly, same result:

```
python update_resources.py --lang                           
> pyuic6 stashdialog.ui
Traceback (most recent call last):
  File "/home/fabiano/BARRACUDA/GIT/gitfourchette/update_resources.py", line 309, in <module>
    compileUi(args.uic, fullpath, F"{root}/ui_{basename}.py", force=args.force, cleanupOutput=not args.no_uic_cleanup)
  File "/home/fabiano/BARRACUDA/GIT/gitfourchette/update_resources.py", line 159, in compileUi
    result = call(uic, os.path.basename(uiPath), cwd=os.path.dirname(uiPath))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/fabiano/BARRACUDA/GIT/gitfourchette/update_resources.py", line 77, in call
    return subprocess.run(args, encoding='utf-8', capture_output=capture_output, check=check, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.12/subprocess.py", line 1955, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'pyuic6'

```

If I try to install pyside6 through `sudo pacman -S pyside6` Konsole tells me that `pyside6-6.8.1-1` is updated but will be reinstalled... any idea what the problem might be?

EDIT: Ah I forgot. If I execute `pip install PySide6` as you suggested, it throws another error with 3 possible solutions:
```

pip install PySide6                                        
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try 'pacman -S
    python-xyz', where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Arch-packaged Python package,
    create a virtual environment using 'python -m venv path/to/venv'.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip.
    
    If you wish to install a non-Arch packaged Python application,
    it may be easiest to use 'pipx install xyz', which will manage a
    virtual environment for you. Make sure you have python-pipx
    installed via pacman.
    
note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
```


I've read [here](https://stackoverflow.com/questions/75602063/pip-install-r-requirements-txt-is-failing-this-environment-is-externally-mana) that the second solution might be what I need, but it sounds very complicated for such a simple task like updating a text file, so I wanted to ask you for confirmation first.

**Reply by jorio** [reply]
Thank you for your response! On Arch, I use AUR package [pyside6-tools-wrappers](https://aur.archlinux.org/packages/pyside6-tools-wrappers) to be able to call `pyside6-lupdate` (instead of pyside_tool.py), I suppose it'll work on Manjaro as well.

Re: the second error – I forgot to mention that `update_resources.py` requires both PySide6 **and PyQt6** to ensure that GitFourchette is compatible with both. At runtime it only needs either of these, but for development we need both. The error you're getting is because the script fails to call `pyuic6`, which is part of PyQt6; you can install `python-pyqt6` to solve this.

Now I realize that installing all these development packages isn’t super convenient just to translate the app, so I'm actually trialing [Weblate](https://weblate.org) to make localization easier. I'll let you know how it goes!

---

PS: All the commands in the guide aren't really necessary to work on the translation itself. If you don't want to set up the dev tooling, you can simply open [gitfourchette_it.ts](https://github.com/user-attachments/files/18118865/gitfourchette_it.ts.gz) in Qt Linguist and work on it there. (Qt Linguist is available as command `linguist6` in Arch package `qt6-tools`, which you might already have.)

**Comment by jorio** [comment, marked as answer]
Hi @FabianoIlCapo – You spurred me to streamline the translation process.

You now have the choice to use Weblate (a convenient web translation platform), or POEdit on your own machine (no more dev tooling to set up):

- **Via Weblate (recommended):** Here's [GitFourchette on Weblate](https://hosted.weblate.org/projects/gitfourchette/gitfourchette/), and here's a blank [Italian localization](https://hosted.weblate.org/translate/gitfourchette/gitfourchette/it) ready for you. Just translate at your own pace and Weblate will take care of the rest.

- **Offline:** You don't need to install the dev toolchain anymore; the only thing you need is [POEdit](https://poedit.net). Open `gitfourchette/assets/lang/gitfourchette.pot` in POEdit, select Italian, and start translating. I've [updated the instructions](https://github.com/jorio/gitfourchette/blob/master/gitfourchette/assets/lang/README.md) if you need more details.

Sorry for this change – Since you're the first person to translate GitFourchette, I figured it was high time to simplify the process.  If you had already made some progress with the previous setup (Qt Linguist), that work isn't lost – just send me your `.ts` file and I'll initialize the new Italian file with it. Thank you!

**Reply by Flaconia** [reply]
Awesome! Glad to be the spark that ignited the change :D

No progress yet, I was going to ask you how long the update process would take... not long at all!
I've tried Weblate in the past, time to test POEdit then and use the faster and most intuitive one.

I'll let you know how it goes

**Reply by Flaconia** [reply]
Hey @jorio 
I'm afraid I lost that first place as a translator. 🥉
It took a while but I finally started the translation on Weblate.

I wanted to ask you: will there be a way to test the translation locally before submitting any official change?
Cheers

**Reply by jorio** [reply]
Hey @FabianoIlCapo! No worries, I'm glad that Italian is joining the party. Grazie!

Here's how to test your translation locally:

1. Download the gettext MO file for your language from Weblate ([direct link for Italian .mo file](https://hosted.weblate.org/download/gitfourchette/gitfourchette/it/?format=mo))
2. Rename `gitfourchette-gitfourchette-it.mo` to just `it.mo`
3. Drop the file into `gitfourchette/assets/lang` beside the existing translations (`fr.mo`, `zh_Hans.mo`, etc.)
4. In GitFourchette, go to File > Settings > Language and Italian should show up.

If you want to preview your changes in the Flatpak version, to do step 3, you can drop the .mo file into `/var/lib/flatpak/app/org.gitfourchette.gitfourchette/current/active/files/lib/python/site-packages/gitfourchette/assets/lang/` (you'll need root access to modify this folder).
